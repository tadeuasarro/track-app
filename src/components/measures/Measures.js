import { useState } from 'react';
import Loading from '../loading/Loading';
import Footer from '../footer/Footer';
import './measures.css';
import createExpenditure from '../../api/createExpenditure';

const Measures = () => {
  const [state, setState] = useState({
    pending: false,
    error: false,
  });

  const handleClick = () => {
    const { expense } = document.getElementById('form-category');
    const { description } = document.getElementById('form-description');
    const { date } = document.getElementById('form-date');
    const { value } = document.getElementById('form-value');

    setState({
      ...state,
      pending: true,
    });

    createExpenditure({
      expense,
      description,
      date,
      value,
    });
  };

  if (state.pending) return <Loading />;

  return (
    <div>
      <div className="measures-container">
        <h3>Create expense</h3>
        <form className="measures-form">
          <input id="form-category" className="measures-form-input" placeholder="Category" />
          <input id="form-description" className="measures-form-input" placeholder="Description" />
          <input id="form-date" className="measures-form-input" placeholder="Date yyyy/mm/dd" />
          <input id="form-value" className="measures-form-input" placeholder="Value" />
          <button onClick={() => handleClick()} className="measures-form-button" type="button">Submit</button>
        </form>
      </div>
      <Footer />
    </div>
  );
};

export default Measures;
