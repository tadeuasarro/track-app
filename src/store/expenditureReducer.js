import { SET_EXPENDITURES } from '../actions/expenditure';
import filterExpenditures from '../helpers/filterExpenditures';

const initialState = {
  expenditures: [],
  summary: false,
};

const expenditureReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_EXPENDITURES:
      return {
        expenditures: action.payload,
        summary: filterExpenditures(action.payload),
      };
    default:
      return state;
  }
};

export default expenditureReducer;
