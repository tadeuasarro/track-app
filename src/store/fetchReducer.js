import { SET_LOADING, SET_ERRORS } from '../actions/fetch';

const initialState = {
  loading: false,
  errors: false,
};

const fetchReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_LOADING:
      return {
        loading: true,
        errors: false,
      };
    case SET_ERRORS:
      return {
        loading: false,
        errors: action.payload,
      };
    default:
      return state;
  }
};

export default fetchReducer;
