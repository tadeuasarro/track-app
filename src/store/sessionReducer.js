import { SET_CURRENT_USER, UPDATE_USER_TARGET } from '../actions/session';

const initialState = {
  username: false,
  id: false,
  target: false,
  token: false,
};

const sessionReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_CURRENT_USER:
      return {
        username: action.payload.user.username,
        id: action.payload.user.id,
        target: action.payload.user.target,
        token: action.payload.token,
      };
    case UPDATE_USER_TARGET:
      return {
        ...state,
        target: action.payload,
      };
    default:
      return state;
  }
};

export default sessionReducer;
