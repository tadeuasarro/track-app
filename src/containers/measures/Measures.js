/* eslint-disable prefer-destructuring */
/* eslint-disable camelcase */
import { useSelector, useDispatch } from 'react-redux';
import { useState } from 'react';
import Footer from '../../components/footer/Footer';
import Error from '../../components/error/Error';
import './measures.css';
import createExpenditure from '../../api/createExpenditure';
import updateExpenditures from '../../helpers/updateExpenditures';
import { setErrors, setLoading } from '../../actions/fetch';

const Measures = () => {
  const { session, fetch } = useSelector(state => state);
  const dispatch = useDispatch();

  const [state, setState] = useState({});

  const handleChange = e => {
    const { name, value } = e.target;
    setState({
      ...state,
      [name]: value,
    });
  };

  const handleClick = async () => {
    dispatch(setLoading());

    const {
      category,
      description,
      date,
      value,
    } = state;

    const res = await createExpenditure({
      category,
      description,
      date,
      value,
      user_id: session.id,
    });

    dispatch(setErrors(res.error));

    if (!res.error) {
      updateExpenditures(session.id, dispatch);
    }
  };

  const errorObj = (!fetch.errors ? {} : fetch.errors);

  if (!session.target) {
    return (
      <div>
        <div className="require-target">
          <h3>Create expense</h3>
          <p>Set up a target before submitting expenses entries.</p>
          <div className="target-instruction">
            <span>Click here! </span>
            <i className="fas fa-arrow-down" />
          </div>
        </div>
        <Footer />
      </div>
    );
  }

  return (
    <div>
      <div className="measures-container">
        <h3>Create expense</h3>
        <form className="measures-form">
          <select onChange={handleChange} name="category" className="measures-form-input">
            <option value="">Category</option>
            <option value="1">Education</option>
            <option value="2">Grocery</option>
            <option value="3">Health</option>
            <option value="4">Leisure</option>
            <option value="5">Living</option>
            <option value="6">Transport</option>
          </select>
          <Error error={errorObj.category} />
          <input onChange={handleChange} name="description" className="measures-form-input" placeholder="Description" />
          <Error error={errorObj.description} />
          <input onChange={handleChange} name="date" className="measures-form-input" type="date" placeholder="Date yyyy/mm/dd" />
          <Error error={errorObj.date} />
          <input onChange={handleChange} name="value" className="measures-form-input" placeholder="Value" />
          <Error error={errorObj.value} />
          <button onClick={() => handleClick()} className="measures-form-button" type="button">Submit</button>
        </form>
      </div>
      <Footer />
    </div>
  );
};

export default Measures;
