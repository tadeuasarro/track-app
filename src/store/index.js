import { combineReducers, createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import sessionReducer from './sessionReducer';
import expenditureReducer from './expenditureReducer';
import fetchReducer from './fetchReducer';

const store = createStore(
  combineReducers({
    session: sessionReducer,
    expenditure: expenditureReducer,
    fetch: fetchReducer,
  }),
  applyMiddleware(thunk),
);

export default store;
