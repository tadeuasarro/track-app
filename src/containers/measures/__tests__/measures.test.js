import { render, screen, act } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../../../store/index';
import Measures from '../Measures';

describe('Rendering component', () => {
  it('creates an App component', () => {
    act(() => {
      render(
        <BrowserRouter>
          <Provider store={store}>
            <Measures />
          </Provider>
        </BrowserRouter>,
      );
    });
    // eslint-disable-next-line
    const element = screen.getByText('Create expense');
  });
});

describe('Display', () => {
  it('renders correctly', () => {
    const comp = renderer.create(
      <BrowserRouter>
        <Provider store={store}>
          <Measures />
        </Provider>
      </BrowserRouter>,
    ).toJSON();
    expect(comp).toMatchSnapshot();
  });
});
