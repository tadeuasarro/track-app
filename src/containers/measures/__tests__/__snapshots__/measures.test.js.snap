// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display renders correctly 1`] = `
<div>
  <div
    className="require-target"
  >
    <h3>
      Create expense
    </h3>
    <p>
      Set up a target before submitting expenses entries.
    </p>
    <div
      className="target-instruction"
    >
      <span>
        Click here! 
      </span>
      <i
        className="fas fa-arrow-down"
      />
    </div>
  </div>
  <div
    className="footer-container"
  >
    <a
      className="footer-link"
      href="/measures"
      onClick={[Function]}
    >
      <i
        className="fas fa-balance-scale-right"
      />
    </a>
    <a
      className="footer-link"
      href="/"
      onClick={[Function]}
    >
      <i
        className="fas fa-bars"
      />
    </a>
    <a
      className="footer-link"
      href="/progress"
      onClick={[Function]}
    >
      <i
        className="fas fa-tasks"
      />
    </a>
    <a
      className="footer-link"
      href="/more"
      onClick={[Function]}
    >
      <i
        className="fas fa-info"
      />
    </a>
  </div>
</div>
`;
